ops/log.py

```diff
@@ -53,6 +55,20 @@ def setup_root_logging(
         debug: if True, write logs to stderr as well as to juju-log.
         exc_stderr: if True, write uncaught exceptions to stderr as well as to juju-log.
     """
+    logging.captureWarnings(True)
+
+    def custom_warning_formatter(
+        message: typing.Union[str, Warning],
+        category: typing.Type[Warning],
+        filename: str,
+        lineno: int,
+        line: typing.Optional[str] = None,
+    ) -> str:
+        module_name = os.path.basename(filename)
```

Comment:
Super-minor: I thought it was odd to `import os` at the top.
Other options:
- `filename.rsplit("/", 1)[-1]`
- `from os.path import basename`
- log the full path, because intermediate elements can be useful, e.g. `ops/__init__.py` vs `requests/__init__.py`

Reply:
I don't think we should do `from os.path import basename`. If we end up using something else from `os` in the future then we add more and more names to the local namespace, rather than just `os` and having `os` things nicely contained in there :).

Doing a split on a filename is bad practice - it should at least be using `os.sep` and not `/` (and then you're importing `os` again). `os` is imported on startup, so there's almost no cost to importing it here. The modern approach would be for these to be `pathlib.Path` objects and then to do `filename.name`, but this is older code so we get a string not a `Path` and converting it to a `Path` just to get the name seems overkill.

A good point about clashing names (although I think it would be less common than you'd get elsewhere since we don't use many packages and charm libs are single files). I had suggested that we shouldn't log the full path because debug-log is super noisy already and this will make it even more noisy, and a chunk of the initial path is not useful, e.g.: `/var/lib/juju/agents/unit-[app name]-[unit number]/charm/src` for charm code (and the packages path for third-party and ops warnings, although we're more concerned with charm warnings in this situation).

We could try to intelligently strip off the part that's not of interest, but that's possibly over-complicating this, particularly as an initial version. So probably better to indeed just go back to the full filename.

Reply:
I wish we had `__qualname__`, but alas! The warnings module doesn't give us that.